package/MDAnalysis/analysis/pca.py

```diff
@@ -0,0 +1,224 @@
+    variance : array (n_components, )
+        The raw variance explained by each eigenvector of the covariance
+        matrix.
+    explained_variance : array, (n_components, )
+        Percentage of variance explained by each of the selected components.
+        If a subset of components is not chosen then all components are stored
+        and the sum of explained variances is equal to 1.0.
+
+    Methods
+    -------
+    fit(traj=None, select='All', start=None, stop=None, step=None)
+        Find the principal components of selected atoms in a subset of the
+        trajectory.
+    transform(traj=None, n_components=None)
+        Take the original trajectory and project it onto the principal
+        components.
+    inverse_tranform(pca_space)
+        Take a pca_space and map it back onto the trajectory used to create it.
+    """
+
+    def __init__(self, atomgroup, demean=True, reference=None,
+                 n_components=None, **kwargs):
+        """
+        Parameters
+        ----------
+        atomgroup: MDAnalysis atomgroup
+            AtomGroup to be used for PCA.
+        n_components : int, optional
+            The number of principal components to be saved, default saves
+            all principal components, Default: -1
+        start : int, optional
+            First frame of trajectory to use for generation
+            of covariance matrix, Default: None
+        stop : int, optional
+            Last frame of trajectory to use for generation
+            of covariance matrix, Default: None
+        step : int, optional
+            Step between frames of trajectory to use for generation
+            of covariance matrix, Default: None
+        """
+        super(PCA, self).__init__(atomgroup.universe.trajectory,
+                                  **kwargs)
+        self._u = atomgroup.universe
+        self.start = kwargs.get('start')
+        self.stop = kwargs.get('stop')
+        self.step = kwargs.get('step')
+
+        if demean:
+            if frame is None:
+                self.reference = self._u
+            else:
+                # change u to traj number for fitting
+                self._u.trajectory[frame]
+                self.reference = self._u
+            aligned = align.AlignTraj(self.reference, self._u).run()
+            # need help here
+            self._u = mda.Universe(aligned.filename,)
+            atomgroup.universe = self._u
+
+        self._atoms = atomgroup
+        self.n_components = n_components
+        self._n_atoms = self._atoms.n_atoms
+
+    def _prepare(self):
+        n_dim = self._n_atoms * 3
+        self.cov = np.zeros((n_dim, n_dim))
+        self.mean = np.zeros(n_dim,)
+
+        for i, ts in enumerate(self._u.trajectory[self.start:self.stop:self.step]):
+            x = self._atoms.positions.ravel()
+            self.mean += x
+
+        self.mean /= self.n_frames
+
+    def _single_frame(self):
+        x = self._atoms.positions.ravel()
+        x -= self.mean
```

Comment:
Thanks. I'll get on it in the morning.
On Tue, Jul 26, 2016 at 12:49 AM Max Linke <notifications@github.com> wrote:

> In package/MDAnalysis/analysis/pca.py
> <https://github.com/MDAnalysis/mdanalysis/pull/896#discussion_r72202890>:
>
> > +        self._n_atoms = self._atoms.n_atoms
> > +
> > +    def _prepare(self):
> > +        n_dim = self._n_atoms * 3
> > +        self.cov = np.zeros((n_dim, n_dim))
> > +        self.mean = np.zeros(n_dim,)
> > +
> > +        for i, ts in enumerate(self._u.trajectory[self.start:self.stop:self.
>
> step]):
> > +            x = self._atoms.positions.ravel()
> > +            self.mean += x
> > +
> > +        self.mean /= self.n_frames
> > +
> > +    def _single_frame(self):
> > +        x = self._atoms.positions.ravel()
> > +        x -= self.mean
>
> This is where the mean substraction has to be replaced with an alignment.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/896/files/9274b6f951361b2c42f5d83fc06c6f001aba3916#r72202890>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKcARnes7wR7r9n87AW7tVctVWRen_bnks5qZbwBgaJpZM4JIXwX>
> .
>
-- 
Have a wonderful day,
John J. Detlefs